source/reduce/structured_loop_to_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,406 @@
+  // The SPIR-V spec says that to be part of a loop or selection, a block must
+  // at a minimum be dominated by the header of the construct.  Check that we
+  // only get here when that's the case.
+
+  auto dominator_analysis = context_->GetDominatorAnalysis(enclosing_function_);
+  assert(dominator_analysis->Dominates(selection_construct_header->id(),
+                                       block_id));
+  auto merge_inst = selection_construct_header->GetMergeInst();
+  assert(merge_inst);
+
+  // Next, the spec says that to be part of the construct the block cannot be
+  // dominated by the merge block of the construct.
+  auto merge_block_id = merge_inst->GetSingleWordOperand(kMergeNodeIndex);
+  if (dominator_analysis->Dominates(merge_block_id, block_id)) {
+    // The block is dominated by the construct's merge block.  Whether a loop or
+    // selection, the block is not part of the construct.
+    return false;
+  }
+
+  if (merge_inst->opcode() == SpvOpLoopMerge) {
+    // In the case of a loop we have the further requirement that, to be in
+    // the construct, the block must not be dominated by the loop's continue
+    // target.
+    if (dominator_analysis->Dominates(
+            merge_inst->GetSingleWordOperand(kContinueNodeIndex), block_id)) {
+      return false;
+    }
+  }
+
+  // The SPIR-V spec's definition of structured control flow is unnatural for
+  // various scenarios where loop breaks and continues are nested inside
+  // conditionals.  To account for this in the short term, we impose the
+  // additional restriction that the block must appear, in program order,
+  // before the construct's merge block in order to be part of the construct.
+  //
+  // TODO: This is a workaround that should be replaced in due course with a
+  // more precise check if the rules for structured control flow are refined.
+  for (auto& bb : *enclosing_function_) {
+    if (bb.id() == block_id) {
+      // We find block_id before the merge block, so we decide that block_id
+      // is part of the structured control flow construct.
+      return true;
+    }
+    if (bb.id() == merge_block_id) {
+      // We find the merge block before block_id.  This suggests that block_id
+      // should not really be part of the construct.
+      return false;
+    }
+  }
+  return true;
+}
+
+void StructuredLoopToSelectionReductionOpportunity::RedirectEdge(
+    uint32_t source_id, uint32_t original_target_id, uint32_t new_target_id) {
+  assert(source_id != original_target_id);
+  assert(source_id != new_target_id);
+  assert(original_target_id != new_target_id);
+
+  // Redirect the edge; depends on what kind of branch instruction is involved.
+  auto terminator = context_->cfg()->block(source_id)->terminator();
+  if (terminator->opcode() == SpvOpBranch) {
+    assert(terminator->GetSingleWordOperand(0) == original_target_id);
+    terminator->SetOperand(0, {new_target_id});
+  } else {
+    assert(terminator->opcode() == SpvOpBranchConditional);
```

Comment:
What about OpSwitch?

Reply:
The RedirectEdge function is specifically for redirecting edges that leave a loop header, so the only choices for the terminator are OpBranch or OpBranchConditional (OpSwitch isn't a legal terminator for a loop header).  I've added a comment to clarify that these are the only choices.

Reply:
Apologies; I was thinking about the wrong bit of code.  Unresolving this so I can think further.

Reply:
Added support (and tests) for OpSwitch.